src/storage/v3/shard_rsm.cpp

```diff
@@ -934,12 +935,115 @@ msgs::ReadResponses ShardRsm::HandleRead(msgs::ScanVerticesRequest &&req) {
+      std::transform(edge_properties.begin(), edge_properties.end(), std::back_inserter(ret),
+                     [&edge](const auto &edge_prop) {
+                       return FromPropertyValueToValue(edge.GetProperty(edge_prop, View::OLD).GetValue());
+                     });
+
+      return ret;
+    };
+  }
+
+  // Functions to select connecting edges based on uniquness
+  using EdgeAccessors = std::vector<memgraph::storage::v3::EdgeAccessor>;
+  std::function<EdgeAccessors(EdgeAccessors &&, memgraph::msgs::EdgeDirection)> maybe_filter_based_on_edge_uniquness;
+  if (req.only_unique_neighbor_rows) {
+    maybe_filter_based_on_edge_uniquness = [](EdgeAccessors &&edges,
+                                              memgraph::msgs::EdgeDirection edge_direction) -> EdgeAccessors {
+      std::function<bool(std::unordered_set<const storage::v3::VertexId *> &,
+                         const memgraph::storage::v3::EdgeAccessor &)>
+          check_and_insert_unique_other_vertex;
+      switch (edge_direction) {
+        case memgraph::msgs::EdgeDirection::OUT: {
+          check_and_insert_unique_other_vertex = [](std::unordered_set<const storage::v3::VertexId *> &other_vertex_set,
+                                                    const memgraph::storage::v3::EdgeAccessor &edge_acc) {
+            auto [it, insertion_happened] = other_vertex_set.insert(&edge_acc.ToVertex());
+            return insertion_happened;
+          };
+          break;
+        }
+        case memgraph::msgs::EdgeDirection::IN: {
+          check_and_insert_unique_other_vertex = [](std::unordered_set<const storage::v3::VertexId *> &other_vertex_set,
+                                                    const memgraph::storage::v3::EdgeAccessor &edge_acc) {
+            auto [it, insertion_happened] = other_vertex_set.insert(&edge_acc.FromVertex());
+            return insertion_happened;
+          };
+          break;
+        }
+        case memgraph::msgs::EdgeDirection::BOTH:
+          MG_ASSERT(false,
+                    "This is should never happen, memgraph::msgs::EdgeDirection::BOTH should not be passed here.");
+      }
+
+      EdgeAccessors ret;
+      std::unordered_set<const storage::v3::VertexId *> other_vertex_set;
+
+      for (const auto &edge : edges) {
+        if (check_and_insert_unique_other_vertex(other_vertex_set, edge)) {
+          ret.emplace_back(edge);
+        }
+      }
+
+      return ret;
+    };
+  } else {
+    maybe_filter_based_on_edge_uniquness =
+        [](EdgeAccessors &&edges, memgraph::msgs::EdgeDirection /*edge_direction*/) -> EdgeAccessors { return edges; };
+  }
+
   auto acc = shard_->Access(req.transaction_id);
   bool action_successful = true;
 
   std::vector<msgs::ExpandOneResultRow> results;
 
   for (auto &src_vertex : req.src_vertices) {
+    // Get Vertex acc
+    auto src_vertex_acc_opt = acc.FindVertex(ConvertPropertyVector((src_vertex.second)), View::OLD);
```

Comment:
```suggestion
    auto src_vertex_acc_opt = acc.FindVertex(ConvertPropertyVector((src_vertex.second)), View::NEW);
```

In general the request should contain the view (not necessary to add it right now), but let's go with the `NEW` view, that eliminates one possible failure during debugging.